### PR TITLE
Password policy

### PR DESCRIPTION
**Pull request by linoman**
**What is this feature?**

This implements an opinionated password function which strengthens the password requirements for Grafana basic auth users.

**Why do we need this feature?**

Full description at [epic issue](https://github.com/grafana/identity-access-team/issues/562)

**Who is this feature for?**

Basic auth users

**Which issue(s) does this PR fix?**:

Fixes #https://github.com/grafana/identity-access-team/issues/563

**Special notes for your reviewer:**

Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

### CODE REVIEW

File: pkg/services/password/password.go
Comment (kalleep) [inline pkg/services/password/password.go:1]: This is not really a service and I don't see a case where it should be. Probably put this in the util package?

File: pkg/api/password.go
Comment (kalleep) [inline pkg/api/password.go:103]: Not sure we need a feature toggle for this one. If the default value for for this config is false this will not be used and `ValidatePassword` will fallback to `.IsWeak` anyway.

File: pkg/services/password/password.go
Comment (linoman) [reply pkg/services/password/password.go:1]: There are a couple of things to consider for this decision.

1. The password validation was done at the user model. Decoupling from that adds flexibility in the future. Moving this static function to utils will achieve the same goal.
2. The password validation is hard-coded. But creating a service for this gives us room for future improvement.

These 2 points were the main reason I decided to use a new service, rather than adding it to utils.

@kalleep do you have a strong preference to move this to the `utils` package?

File: pkg/api/password.go
Comment (linoman) [reply pkg/api/password.go:103]: All new features have to be covered under a feature toggle. There are a lot of [benefits](https://github.com/grafana/hosted-grafana/wiki/All-Things-Feature-Toggles#why-feature-toggles) from having this feature under a toggle and it doesn't hurt to protect the implementation. Once this is approved, the feature toggle will be removed.

File: pkg/services/password/password.go
Comment (linoman) [reply pkg/services/password/password.go:1]: As suggested by @kalleep I'll be moving this to its own package under the `user` package.

File: pkg/api/password.go
Comment (kalleep) [reply pkg/api/password.go:103]: Sorry I don't think I was clear enough. I don't see the need for a feature toggle at al. The config option already is a toggle and is false by default

File: conf/defaults.ini
Comment (mgyongyosi) [inline conf/defaults.ini:815]: Please add a short description of the setting.

File: pkg/setting/setting.go
Comment (mgyongyosi) [inline pkg/setting/setting.go:1595]: Please either update here or in the `defaults.ini` and `sample.ini`, because the `ini` files contain `password_policy` as a key.

File: conf/defaults.ini
Comment (linoman) [reply conf/defaults.ini:822]: Updated as requested

File: pkg/setting/setting.go
Comment (linoman) [reply pkg/setting/setting.go]: This was a mistake by my part, I have fixed this

File: pkg/api/password.go
Comment (linoman) [reply pkg/api/password.go]: I removed the feature toggle

File: pkg/services/user/password.go
Comment (kalleep) [inline pkg/services/user/password.go:18]:
```suggestion
// ValidatePassword checks if a new password meets the required criteria based on the given configuration.
// If BasicAuthStrongPasswordPolicy is disabled, it only checks for password length.
// Otherwise, it ensures the password meets the minimum length requirement and contains at least one uppercase letter,
// one lowercase letter, one number, and one symbol.
```

File: pkg/services/user/password.go
Comment (kalleep) [inline pkg/services/user/password.go:19]: I think we should get rid of `Password` and just use string. Then move just recreate `IsWeek` but as a non exported function here.

Comment (Jguer) [reply pkg/services/user/password.go:19]: I think we don't need to fall into [Primitive Obsession](https://refactoring.guru/smells/primitive-obsession).

I think we just need to change

```go
userPassword := user.Password(newPassword)
	if err := user.ValidatePassword(userPassword, runner.Cfg); err != nil {
		return fmt.Errorf("new password is too short")
	}
```

to 

```go
userPassword, err := user.NewPassword(newPassword)
...
```

Make validation part of the type and not an afterthought

Comment (Jguer) [reply pkg/services/user/password.go:19]: this follows more a pattern of Parsing, not validating

https://lexi-lambda.github.io/blog/2019/11/05/parse-don-t-validate/

Comment (kalleep) [reply pkg/services/user/password.go:19]: Yep that is also a good suggestion

Comment (kalleep) [reply pkg/services/user/password.go:19]: Then you ideally would also update the type in models to not use string but password type instead

File: pkg/services/user/model.go
Comment (Jguer) [inline pkg/services/user/model.go:286]: https://github.com/grafana/grafana/pull/82268#discussion_r1489125227

Lino can you take a look at this?

File: pkg/services/user/password.go
Comment (linoman) [reply pkg/services/user/password.go]: I migrated all he password properties from string type to user.Password type. I'm working on the tests

**Review by mgyongyosi — APPROVED** [review]
Seems to be fine, but let's wait for the re-review from @Jguer and @kalleep.

File: pkg/services/authn/clients/grafana.go
Comment (Jguer) [inline pkg/services/authn/clients/grafana.go:115]:
```suggestion
func comparePassword(password, salt string, hash string) bool {
```

hash is a string

**Review by Jguer — APPROVED** [review]
all good once the hash type is fixed

File: pkg/services/authn/clients/grafana.go
Comment (linoman) [reply pkg/services/authn/clients/grafana.go]: changed to string

**Pull request merged**
